File: src/generator/src/Generator.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.io.LineNumberReader;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Random;
import java.util.TreeMap;
import java.util.TreeSet;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/*
From the distribution of brights (biased to higher Integrated Intensity sequences) 
REPEAT:
1) sample number of dark/bright motifs in parts 1,2,3  
2) average stickiness in parts 1,2,3
// so far we have a "recipe" for a sequence next we use motifs
3) sample motifs according to numbers in step 1) and such we satisfy stickiness from step 2)
REPEAT:
   4) fill in the gaps (multiple times) randomly
   5) ensure the generated sequence is different from our current population
   6) run through classifier and accept if Prob(bright) > Some threshold
*/

public class Generator {
	public static HashMap<String,Double> BASE_STICKY = null;
	static{
		BASE_STICKY = new HashMap<String, Double>();
		BASE_STICKY.put("A", 0.5);
		BASE_STICKY.put("T", 0.0);
		BASE_STICKY.put("C", 1.0);
		BASE_STICKY.put("G", 1.0);
	}
	
	private Random r = new Random(System.currentTimeMillis());
	private int n = 0, nreg = 0;
	private String[] seqs = null;
	private int[][] bcnt;
	private double[][] bsize;
	private double[][] bsticky;
	private int[][] dcnt;
	private String[] classes;
	private ArrayList<Motif> motifs;
	private double[] intensity;
	
	// distributions for sampling
	private ArrayList<TreeMap<Double,Integer>> distBnum = new ArrayList<TreeMap<Double, Integer>>();
	private ArrayList<TreeMap<Double,Integer>> distDnum = new ArrayList<TreeMap<Double, Integer>>();
	private ArrayList<TreeMap<Double,Double>> distSticky = new ArrayList<TreeMap<Double, Double>>();
	private ArrayList<TreeMap<Double,Motif>> distMotif = new ArrayList<TreeMap<Double, Motif>>();
	
	public Generator(String inSeqFn) throws IOException{
		readSequences(inSeqFn);
		computeDists("bright");
	}

	private ArrayList<String> generate(Integer numNew) {
		int cnt = 0;
		int atMostNgaps = 1;
		ArrayList<String> generated = new ArrayList<String>();
		while (cnt < numNew) {
			String seq = "__________";
			seq = placeMotifs(seq, atMostNgaps);
			seq = fillGapsBasedOnStickyness(seq);
			generated.add(seq);
			cnt++;
		}
		return generated;
	}

	private String fillGapsBasedOnStickyness(String seq) {
		// Fill in gaps according to sticky[]
		int ind = -1;
		while ((ind = seq.indexOf('_')) >= 0) {
			ArrayList<Integer> regs = getRegions(ind, seq.length(), nreg);
			TreeMap<String,Double> base2score = new TreeMap<String, Double>();
			double totScore = 0.0;
			for (String base: BASE_STICKY.keySet()) base2score.put(base, 0.0);
			for (Integer reg: regs) {
				int s = reg*(seq.length()/nreg);
				int e = (reg+1)*(seq.length()/nreg) + seq.length()%nreg;
				String str = seq.substring(s,e);
				for (String base: base2score.keySet()) {
					int replace = ind-s;
					assert(str.charAt(replace)=='_');
					String strnew = str.substring(0,replace) + 
							base + str.substring(replace+1,str.length());
					Double avgSticky = computeAvgStickyness(strnew);
					double sc = 0;
					if (null == distSticky.get(reg).floorKey(avgSticky)) {
						sc = distSticky.get(reg).firstKey();
					} else {
						sc = ((null==distSticky.get(reg).ceilingKey(avgSticky))
								?distSticky.get(reg).lastKey()
								:distSticky.get(reg).ceilingKey(avgSticky)) - 
							((null==distSticky.get(reg).floorKey(avgSticky - 0.000000001))?0:
								distSticky.get(reg).floorKey(avgSticky - 0.000000001));
					}
					base2score.put(base, base2score.get(base) + sc);
					totScore += sc;
				}
			}
			TreeMap<Double,String> prob2base = new TreeMap<Double, String>();
			double acc = 0.0;
			for(String base: base2score.keySet()) {
				prob2base.put(base2score.get(base)/totScore + acc, base);
				acc += base2score.get(base)/totScore;
			}
			String base = prob2base.get(prob2base.ceilingKey(r.nextDouble()));
			seq = seq.substring(0,ind) + base + seq.substring(ind+1,seq.length());
		}
		return seq;
	}

	private String placeMotifs(String seq, int atMostNgaps) {
		int[] pos = getShuffledIndices();
		int numGaps = seq.length();
		// sample motifs in 1 then 2 and 3 positions
		while (numGaps > atMostNgaps) {
			for (int j=0;j<nreg;j++) {
				int i = pos[j];
				Motif m = distMotif.get(i).get(distMotif.get(i).ceilingKey(r.nextDouble()));
				ArrayList<Integer> possible_pos = getPossiblePositions(m,seq,i);
				if(possible_pos.size()>0) {
					int position = possible_pos.get(r.nextInt(possible_pos.size()));
					seq = insert(m,seq,position);
				}
			}
			numGaps = 0;
			for (int i=0; i < seq.length(); i++)
				if (seq.charAt(i) == '_') 
					numGaps++;
		}
		return seq;
	}

	private static Double computeAvgStickyness(String strnew) {
		int cnt = 0;
		double sticky = 0.0;
		for (int i = 0; i < strnew.length(); i++) {
			if (strnew.charAt(i)!='_') {
				cnt++;
				sticky += BASE_STICKY.get(strnew.substring(i, i+1));
			}
		}
		if (cnt>0) sticky /= cnt;
		return sticky;
	}

	private static ArrayList<Integer> getRegions(int ind, int length, int numreg) {
		int chunk = length / numreg;
		int rem = length % numreg;
		ArrayList<Integer> res = new ArrayList<Integer>();
		for (int pos = 0; pos < numreg; pos++) {
			if(ind>=pos*chunk && ind < (pos+1)*chunk + rem)
				res.add(pos);
		}
		return res;
	}

	private int[] getShuffledIndices() {
		int[] pos = new int[nreg];
		for (int i=0; i<nreg; i++) pos[i] = i;
		for (int i=0; i<10; i++) {
			int a = r.nextInt(nreg);
			int b = r.nextInt(nreg);
			int sub = pos[a];
			pos[a] = pos[b];
			pos[b] = sub;
		}
		return pos;
	}

	private static int editDist(String seq1, String seq2) {
		int d = 0;
		assert(seq1.length() == seq2.length());
		for (int i=0; i < seq1.length(); i++) {
			if(seq1.charAt(i)!='_' && seq2.charAt(i)!='_' 
					&& seq1.charAt(i) != seq2.charAt(i)) d++;
		}
		return d;
	}

	private String insert(Motif m, String seq, int pos) {
		assert(fits(m,seq,pos)); 
		String res = "";
		for (int i=0; i < seq.length(); i++) {
			if (i<pos || i >= pos+m.regex.length()) res += seq.charAt(i);
			else if (seq.charAt(i) == '_') res += m.regex.charAt(i-pos);
			else if (m.regex.charAt(i-pos) == '_') res += seq.charAt(i);
			else {
				assert(seq.charAt(i) == m.regex.charAt(i-pos));
				res += seq.charAt(i);
			}
		}
		return res;
	}

	private ArrayList<Integer> getPossiblePositions(Motif m, String seq, int region) {
		ArrayList<Integer> resArrayList = new ArrayList<Integer>();
		double half = m.regex.length()/2.0;
		for (int st = 0; st < seq.length()-m.regex.length()+1; st++) {
			if (Math.floor((st+half)*nreg*1.0/seq.length()) == region && fits(m, seq, st)) 
				resArrayList.add(st);
		}
		return resArrayList;
	}

	public boolean fits(Motif m, String seq, int position) {
		for (int i=0; i< m.regex.length(); i++) {
			if (m.regex.charAt(i)!='_' && seq.charAt(i+position) != '_' && 
				m.regex.charAt(i) != seq.charAt(i+position)) {
				return false;
			}
		}
		return true;
	}
	
	private static boolean matchInReg(Motif m, String seq, int reg, int numreg) {
		assert(!seq.contains("_"));
		String regex = m.regex.replace("_", "[ACTG]{1}");
		Pattern p = Pattern.compile(regex);
		Matcher matcher = p.matcher(seq);
		while(matcher.find()) {
			double middle = (matcher.start() + matcher.end())/2.0;
			if(Math.floor(middle*numreg*1.0/seq.length()) == reg)
				return true;
		}
		return false;
	}

	// STATS #######################################################
	
	private void plotNeighborStats(ArrayList<String> generated, int maxdist) {
		TreeMap<Double,Integer> stats_close = new TreeMap<Double, Integer>();
		for (int i =0; i < seqs.length; i++) stats_close.put(i*1.0, 0);
		for (String seq: generated) {
			countNeighboringSequences(stats_close, seq, maxdist);
		}
		int tot = 0, b = 0, d = 0;
		for (Double ind: stats_close.keySet()) {
			if("bright".equals(classes[ind.intValue()])) b+= stats_close.get(ind);
			else d+=stats_close.get(ind);
			tot+=stats_close.get(ind);
		}
		Double perc = b*100.0/tot;
		System.out.print( "bright: " + perc.intValue() + "%\n");
		Common.plotHist(stats_close);
	}
	
	private void countNeighboringSequences(
			TreeMap<Double, Integer> stats_close, String seq, int radius) {
		for (int i = 0 ; i < seqs.length; i++) {
			int d = editDist(seqs[i],seq);
			if (d<=radius) {
				stats_close.put(i*1.0, stats_close.get(i*1.0) + 1);
			}
		}
	}
	
	private boolean isClosestKnownWithin(int minD, int maxD, String s) {
		int closestD = s.length(), d = 0;
		assert(s.length() == seqs[0].length());
		for (int i = 0 ; i < seqs.length; i++) {
			d = editDist(seqs[i],s);
			if( d < closestD)
				closestD = d;
		}
		return (minD<=closestD) && (maxD >= closestD);
	}
	
	// DISTRIBUTIONS ###################################################
	
	private void computeDists(String clas) {
		// compute the total II in bright
		double sumII = getTotalIntensity(clas);
		for (int pos = 0; pos < nreg; pos++) {
			distBnum.add(computeDist(clas,sumII,pos,bcnt));
			distDnum.add(computeDist(clas,sumII,pos,dcnt));
			distSticky.add(computeDist(clas, sumII, pos, bsticky));
		}
		distMotif = computeDistMotif(clas);
	}

	private ArrayList<TreeMap<Double,Motif>> computeDistMotif(String clas) {
		ArrayList<TreeMap<Double,Motif>> res = new ArrayList<TreeMap<Double,Motif>>();
		double[] totalII = new double[nreg];
		for (int reg = 0; reg < nreg; reg++) {
			for (Motif m: motifs ) {
				for (int i = 0; i < n; i++) {
					if (classes[i].equals(clas)) {
						if (matchInReg(m,seqs[i],reg,nreg)){
							m.prob[reg] += intensity[i];
							totalII[reg] += intensity[i];
						}
					}
				}
			}
			if (totalII[reg]>0)
				for (Motif m: motifs) 
					m.prob[reg] /= totalII[reg];
			TreeMap<Double, Motif> prob2mot = new TreeMap<Double, Motif>();
			double acc = 0.0;
			for (Motif m: motifs) {
				prob2mot.put(m.prob[reg]+acc, m);
				acc += m.prob[reg];
			}
			res.add(prob2mot);
		}
		return res;
	}

	public double getTotalIntensity(String clas) {
		double sumII = 0.0;
		for (int i = 0; i < n; i++) {
			if (classes[i].equals(clas)) {
				sumII += intensity[i];
			}
		}
		return sumII;
	}

	public TreeMap<Double, Integer> computeDist(String clas, double sumII, int pos, int[][] vals) {
		TreeMap<Integer, Double> m = new TreeMap<Integer, Double>();	
		for (int i = 0; i < n; i++) {
			if (classes[i].equals(clas)) {
				if (!m.containsKey(vals[i][pos])) m.put(vals[i][pos], 0.0);
				m.put(vals[i][pos], m.get(vals[i][pos]) + intensity[i]/sumII);
			}
		}
		TreeMap<Double, Integer> prob2num = new TreeMap<Double, Integer>();
		double acc = 0.0;
		for (Integer num: m.keySet()) {
			prob2num.put(m.get(num)+acc, num);
			acc += m.get(num);
		}
		return prob2num;
	}
	
	public TreeMap<Double, Double> computeDist(String clas, double sumII, int pos, double[][] vals) {
		TreeMap<Double, Double> m = new TreeMap<Double, Double>();	
		for (int i = 0; i < n; i++) {
			if (classes[i].equals(clas)) {
				if (!m.containsKey(vals[i][pos])) m.put(vals[i][pos], 0.0);
				m.put(vals[i][pos], m.get(vals[i][pos]) + intensity[i]/sumII);
			}
		}
		TreeMap<Double, Double> prob2dbl = new TreeMap<Double, Double>();
		double acc = 0.0;
		for (Double num: m.keySet()) {
			prob2dbl.put(m.get(num)+acc, num);
			acc += m.get(num);
		}
		return prob2dbl;
	}
	
	// READ/WRITE ###################################################
	
	private void readSequences(String fn) throws IOException {
		n= getNumLines(fn) -1;
		BufferedReader br = new BufferedReader(new FileReader(fn));
		String line = br.readLine();
		String[] sline = line.split(",");
		int colFeat = 5, colSeq = 1, colII = 4, colBPos = -1, colDPos = -1, 
				colBsize = -1, colBSticky = -1, colClass = -1;
		for (int i = 0; i < sline.length; i++) {
			if (sline[i].equalsIgnoreCase("BrightPos0")) colBPos = i;
			if (sline[i].equalsIgnoreCase("DarkPos0")) colDPos = i;
			if (sline[i].equalsIgnoreCase("BaseSize0")) colBsize = i;
			if (sline[i].equalsIgnoreCase("BaseSticky0")) colBSticky = i;
			if (sline[i].equalsIgnoreCase("class")) colClass = i;
		}
		nreg=colDPos-colBPos;
		motifs = new ArrayList<Motif>();
		for (int i = colFeat; i < colBPos; i++) {
			motifs.add(new Motif(sline[i],nreg)); 
			if (sline[i].contains("_")) 
				motifs.add(new Motif(sline[i].replace("_", ""),nreg)); 
		}
		
		seqs = new String[n];
		bcnt = new int[n][nreg];
		dcnt = new int[n][nreg];
		bsize = new double[n][nreg];
		bsticky = new double[n][nreg];
		classes = new String[n];
		intensity = new double[n];
		
		int nr = 0;
		while((line = br.readLine())!=null) {
			sline = line.split(",");
			seqs[nr] = sline[colSeq];
			classes[nr] = sline[colClass];
			intensity[nr] = Double.parseDouble(sline[colII]);
			for(int i = 0; i < nreg; i++) {
				bcnt[nr][i] = Integer.parseInt(sline[colBPos + i]);
				dcnt[nr][i] = Integer.parseInt(sline[colDPos + i]);
				bsize[nr][i] = Double.parseDouble(sline[colBsize + i]);
				bsticky[nr][i] = Double.parseDouble(sline[colBSticky + i]);
			}	
			nr++;
		}
	}
	
	private int getNumLines(String fn) throws FileNotFoundException, IOException {
		LineNumberReader lnr = new LineNumberReader(new FileReader(fn));
		lnr.skip(Long.MAX_VALUE);
		return lnr.getLineNumber();
	}
	
	private void writeSequences(String fn, ArrayList<String> sequences, int MinD, int MaxD) throws IOException{
		BufferedWriter bw = new BufferedWriter(new FileWriter(fn));
		int id =0;
		TreeSet<String> nseq = new TreeSet<String>();
		for(String s: sequences) nseq.add(s);
		bw.write("Name,Sequence\n");
		for(String s:nseq) {
			if (isClosestKnownWithin(MinD,MaxD,s)) {
				bw.write("10basegen"+id + "," + s + "\n");
				id++;
			}
		}
		System.out.print("Wrote " + id + " new sequences\n");
		bw.close();
	}
	
	// TEST #####################################

	public static void testMatchInRegion() throws IOException {
		String seq = "ATCCCCAAAC";
		Motif m = new Motif("AA_C+",3);
		assert(matchInReg(m, seq, 2, 3));
	}
	
	// MAIN #####################################
	public static void main(String[] args) throws IOException{
//		testMatchInRegion();
//		System.exit(0);
		String inSeqFn = args[0];
		Integer numNew = Integer.parseInt(args[1]);
		int Tmin = Integer.parseInt(args[2]);
		int Tmax = Integer.parseInt(args[3]);
		
		Generator gen = new Generator(inSeqFn);
		ArrayList<String> newseqs = gen.generate(numNew);
		gen.plotNeighborStats(newseqs, 3);
		gen.writeSequences(inSeqFn.substring(0,inSeqFn.indexOf(".")) + "-new.csv", 
				newseqs, Tmin, Tmax);
	}

}

File: start/generator.py
import random, sys, re, math, os, env, csv

# Globals
BASE_STICKY = {"A" : 0.5, "T" : 0.0, "C" : 1.0, "T" : 0.0}
RAND = random.seed()

def getNumLines(fname):
    with open(fname) as f:
        for i, l in enumerate(f):
            pass
    return i + 1

def ceilingKey(d, key):
    ceilK = []
    if key in d:
        return key
    ceilK = [k for k in d if k >= key]
    if ceilK:
        return min(ceilK)
    return 0.0      # TODO is returning 0 here aproblem??
    # return min(k for k in d if k >= key)            

def floorKey(d, key):
    floorK = []
    if key in d:
        return key
    floorK = [k for k in d if k <= key]
    if floorK:    
        return max(floorK)
    return 0.0
    
# Generator 
class Generator:
    def __init__(self, fn):
        self.fileName = fn
        self.n = getNumLines(self.fileName)
        self.motifs = []
        self.seqs = [""]*self.n
        self.distBnum = []
        self.distDnum = []
        self.distSticky = []
        self.distMotif = []
        self.seqValHash = {}

    def getRegions(self, ind, length, numreg):
        chunk = length / numreg
        rem = length % numreg
        res = []
        for pos in range(numreg):
            if ind >= pos*chunk and ind < (pos+1)*chunk + rem:
                res.append(pos)
        return res

    def computeAvgStickyness(self, strnew):
        cnt = 0
        sticky = 0.0
        for i in range(len(strnew)):
            if strnew[i] == "_":
                cnt = cnt +1
                sticky += BASE_STICKY[strnew[i:i+1]]
        if cnt > 0:
            sticky /= cnt
        return sticky
        
    def fillGapsBasedOnStickyness(self, seq):
        try:
            ind = seq.index('_')
        except ValueError:
            return seq
        # print "ind: ",ind
        regs = self.getRegions(ind, len(seq), self.nreg)
        # print "regs: ", regs
        base2score ={}
        totScore = 0.0
        for base in BASE_STICKY:
            base2score[base] = 0.0
        for reg in regs:
            s = reg * (len(seq)/self.nreg)
            e = (reg+1)*(len(seq)/self.nreg) + len(seq)%self.nreg
            str = seq[s:e]
            for base in base2score:
                replace = ind - s
                assert(str[replace] == "_")
                strnew = str[0:replace] + base + str[replace+1:len(str)]
                avgSticky = self.computeAvgStickyness(strnew)
                sc = 0
                if floorKey(self.distSticky[reg], avgSticky) is None:
                    sc = firstKey(self.distSticky[reg])
                else:
                    if ceilingKey(self.distSticky[reg], avgSticky) is None:
                        sc = self.distSticky[reg].keys()[-1]
                    else:
                        if floorKey(self.distSticky[reg], avgSticky - 0.000000001) is None:
                            sub = 0
                        else:
                            sub = floorKey(self.distSticky[reg], avgSticky - 000000001)
                        sc = ceilingKey(self.distSticky[reg], avgSticky) - sub
                base2score[base] = base2score[base] + sc
                totScore += sc
        prob2base = {}
        acc = 0.0
        for base in base2score:
            prob2base[base2score[base]/totScore + acc] = base
            acc += base2score[base] / totScore
        base = prob2base[ceilingKey(prob2base, random.random())]
        # print "base: ",base    
        seq = seq[0:ind] + base + seq[ind+1:len(seq)]
        return seq
    
    def generate(self, numNew):
        cnt = 0
        atMostNGaps = 1
        generated = []
        while cnt < numNew:
            seq = "__________"
            # print "calling placeMotifs", seq, atMostNGaps
            seq = self.placeMotifs(seq, atMostNGaps)
            # print "before :",seq
            seq = self.fillGapsBasedOnStickyness(seq)
            # print "after :", seq, "\n"
            generated.append(seq)
            cnt = cnt + 1
        return generated

    def getPossiblePositions(self, m, seq, region):
        resArrayList = []
        half = len(m.regex) / 2.0
        # print seq, m.regex
        # print "seq before (getpossiblepositions)", seq
        for st in range(len(seq) - len(m.regex) +1):
            if math.floor((st + half)*self.nreg*1.0/len(seq)) == region and self.fits(m, seq, st):
                resArrayList.append(st)
        # print "seq after (getpossiblepositions)", seq
        return resArrayList

    def fits(self, m, seq, position):
        for i in range(len(m.regex)):
            # print "seq before (fits)", seq
            if m.regex[i] is not "_" and seq[i+position] is not "_" and m.regex[i] is not seq[i+position]:
                return False
            # print "seq after (fits)", seq
        return True

    def insert(self, m, seq, pos):
        # print "in insert, seq: ", seq    
        assert(self.fits(m, seq, pos))
        res = ""
        for i in range(len(seq)):
            if i < pos or i >= pos + len(m.regex):
                res += seq[i]
            elif seq[i] == "_":
                res += m.regex[i-pos]
            elif m.regex[i-pos] == "_":
                res += seq[i]
            else:
                assert(seq[i] == m.regex[i-pos])
                res += seq[i]
        return res
        
    def placeMotifs(self, seq, atMostNGaps):
        pos = self.getShuffledIndices()
        numGaps = len(seq)
        # sample motifs in 1 then 2 and 3 positions
        while(numGaps > atMostNGaps):
            for j in range(self.nreg):
                i = pos[j]
                m = self.distMotif[i][ceilingKey(self.distMotif[i], random.random())]
                # print "calling getPossiblePositions", m, seq, i
                possible_pos = self.getPossiblePositions(m, seq, i)
                if len(possible_pos) > 0:
                    position = possible_pos[random.randint(0, len(possible_pos)-1)]
                    seq = self.insert(m, seq, position)
            numGaps = 0
            for i in range(len(seq)):
                if seq[i] == "_":
                    numGaps = numGaps + 1
        return seq 
                    
    def getShuffledIndices(self):
        pos = [0] * self.nreg
        for i in range(self.nreg):
            pos[i] = i
        for i in range(10):
            a = random.randint(0, self.nreg-1)
            b = random.randint(0, self.nreg-1)
            sub = pos[a]
            pos[a] = pos[b]
            pos[b] = sub
        return pos
        
    def makeArrays(self, nreg):
        self.nreg = nreg    
        self.bcnt = [[0 for x in range(self.nreg)] for x in range(self.n)]
        self.dcnt = [[0 for x in range(self.nreg)] for x in range(self.n)]
        self.bsize = [[0 for x in range(self.nreg)] for x in range(self.n)]
        self.bsticky = [[0 for x in range(self.nreg)] for x in range(self.n)]
        self.classes = [""] * self.n
        self.intensity = [0] * self.n 

    def readIntensityFile(self, fn):
        # Read intensity file; Values could also be binary (0/1)
        with open(fn, "r") as csvFile:
            csvreader = csv.reader(csvFile)
            # build Hash of Sequence, Value
            for line in csvreader:
                # line[0] is sequence, line[1] is intensity value
                self.seqValHash[line[0]] = line[1]
        # print self.seqValHash
                
    def readTrainingFile(self):
        # get relevant info. from training file
        with open(self.fileName, "r") as trainingFile:
            line = trainingFile.readline()
            sline = line.split(",")
            colFeat = 1
            colSeq = 0
            # colII = 4
            colBPos = colDPos = colBsize = colBSticky = colClass = -1
            for i in range(0, len(sline)):
                # print sline[i]
                if sline[i].lower() == "brightpos0": colBPos = i
                if sline[i].lower() == "darkpos0": colDPos = i
                if sline[i].lower() == "basesize0": colBsize = i
                if sline[i].lower() == "basesticky0": colBSticky = i
                if sline[i].lower() == "class": colClass = i
            # print colBPos, colDPos, colBsize, colBSticky, colClass
            nreg = colDPos - colBPos
            self.makeArrays(nreg)
            for i in range(colFeat, colBPos):
                self.motifs.append(Motif(sline[i], nreg))
                if "_" in sline[i]:
                    self.motifs.append(Motif(sline[i].replace("_", ""), nreg))
            nr = 0
            # print bcnt
            for line in trainingFile:
                # print sline
                sline = line.split(",")
                self.seqs[nr] = sline[colSeq]
                self.classes[nr] = sline[colClass][:-1] # truncate newline
                # print "seq",self.seqs[nr]
                # exit(0)
                # self.intensity[nr] = float(sline[colII])
                self.intensity[nr] = float(self.seqValHash[self.seqs[nr]])
                for i in range(nreg):
                    self.bcnt[nr][i] = int(sline[colBPos+i])
                    self.dcnt[nr][i] = int(sline[colDPos+i])
                    self.bsize[nr][i] = float(sline[colBsize+i])
                    self.bsticky[nr][i] = float(sline[colBSticky+i])
                nr = nr + 1
    def getTotalIntensity(self, clas):
        sumII = 0.0
        # print self.classes
        for i in range(self.n):
            if self.classes[i] == clas:
                sumII = sumII + self.intensity[i]
        # print sumII
        return sumII
    def computeDist(self, clas, sumII, pos, vals):
        m = {}
        for i in range(self.n):
            if self.classes[i] == clas:
                if vals[i][pos] not in m:
                    m[vals[i][pos]] = 0.0
                m[vals[i][pos]] = m[vals[i][pos]] + self.intensity[i] / sumII
        prob2num = {}
        acc = 0.0
        for num in m:
            prob2num[m[num]+acc] = num
            acc += m[num]
        # print prob2num, "\n"
        return prob2num

    def matchInReg(self, m, seq, reg, numreg):
        assert "_" not in seq
        regex = m.regex.replace("_", "[ACTG]{1}")
        pattern = re.compile(regex)
        for match in pattern.finditer(seq):
            # print match.start(), match.group(), seq, regex, reg, numreg
            middle = (match.start() + len(match.group())) / 2.0
            if math.floor(middle*numreg*1.0/len(seq)) == reg:
                # print "yes"
                return True
        # print "no"
        return False
            
    def computeDistMotif(self, clas):
        res = []
        totalII = [0.0] * self.nreg
        for reg in range(self.nreg):
            for m in self.motifs:
                for i in range(self.n):
                    if self.classes[i] == clas:
                        if self.matchInReg(m, self.seqs[i], reg, self.nreg):
                            m.prob[reg] += self.intensity[i]
                            totalII[reg] += self.intensity[i]
            if totalII[reg] > 0:
                for m in self.motifs:
                    m.prob[reg] /= totalII[reg]
            prob2mot = {}
            acc = 0.0
            for m in self.motifs:
                prob2mot[m.prob[reg]+acc] = m
                acc += m.prob[reg]
            res.append(prob2mot)
        return res            
            
    def computeDists(self,clas):
        # compute total integrated intensity in bright
        sumII = self.getTotalIntensity(clas)
        for pos in range(self.nreg):
            self.distBnum.append(self.computeDist(clas, sumII, pos, self.bcnt))
            self.distDnum.append(self.computeDist(clas, sumII, pos, self.dcnt))
            self.distSticky.append(self.computeDist(clas, sumII, pos, self.bsticky))
        self.distMotif = self.computeDistMotif(clas)

    def editDist(self, seq1, seq2):
        d = 0
        # print seq1, seq2
        # assert(len(seq1) == len(seq2))
        if len(seq1) is not len(seq2):
            return
        for i in range(len(seq1)):
            if seq1[i] is not "_" and seq2[i] is not "_" and seq1[i] is not seq2[i]:
                d = d + 1
        return d
        
    def isClosestKnownWithin(self, minD, maxD, s):
        closestD = len(s)
        d = 0
        assert(len(s) == len(self.seqs[0]))
        for i in range(len(self.seqs)):
            d = self.editDist(self.seqs[i], s)
            if d < closestD:
                closestD = d
        return minD <= closestD  and maxD >= closestD
        
    def writeSequences(self, fileName, sequences, MinD, MaxD):
        ident = 0
        with open(fileName, "w") as outf:
            nseq = set()
            for s in sequences:
                # print "sequence s :",s
                nseq.add(s)
            outf.write("Name,Sequence\n")
            # print nseq    
            for s in nseq:
                # if self.isClosestKnownWithin(MinD, MaxD, s):
                outf.write("10basegen" + str(ident) + "," + s + "\n")
                ident = ident + 1
            print "wrote", ident, "new sequences"
        
    def printALL(self):
        print "printing ALL: \n", \
            "Motifs: ", str(self.motifs), "\n\n", \
            "Seqs: ", str(self.seqs), "\n\n", \
            "bcnt: ", str(self.bcnt), "\n\n", \
            "dcnt: ", str(self.dcnt), "\n\n"
        for m in self.motifs:
            print m.regex
# Motif
class Motif:
    def __init__(self, nm, nreg):    
        self.name = nm
        self.isPos = nm.endswith("+")
        self.regex = nm[0:len(nm)-1]
        self.prob = [0]*nreg
    def printAll(self):
        print "printing ALL: " + self.name, self.isPos, self.regex, self.prob

def doAll(trainingFn, intensityFn, numNew, Tmin, Tmax, useThresholdp, threshList=None):
    if threshList:
        print "threshlist",threshList
    print "intensityFn",intensityFn
    # exit(0)
    outFileName = os.path.join(env._env['NEW_FEATURES_PATH'], trainingFn)
    outFileName = os.path.splitext(outFileName)[0] + "-new.csv"
    print "outFileName",outFileName
    # exit(0)
    infileName = os.path.join(env._env['TRAINING_DATA_PATH'], trainingFn)
    print "infileName",infileName
    # exit(0)
    generator = Generator(infileName) # instantiate object; get num lines in file
    # build in memory hash for Sequence,Value; 
    generator.readIntensityFile(intensityFn) 
    # exit(0)
    generator.readTrainingFile()
    # exit(0)
    generator.computeDists("bright")
    # exit(0)
    newseqs = generator.generate(numNew)
    # print outFileName
    generator.writeSequences(outFileName, newseqs, Tmin, Tmax)



    
if __name__ == "__main__":
    fileName = sys.argv[1]
    numNew = int(sys.argv[2])
    Tmin = int(sys.argv[3])
    Tmax = int(sys.argv[4])    
    generator = Generator(fileName)
    generator.readSequences()
    generator.computeDists("bright")
    newseqs = generator.generate(numNew)
    generator.writeSequences(os.path.splitext(fileName)[0] + "-new.csv", newseqs, Tmin, Tmax)
